<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang xml:lang>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>HW14</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <div style="display: none">
  \[
  %%
  % Add your macros here; they&#39;ll be included in pdf and html output.
  %%

  \newcommand{\R}{\mathbb{R}}    % reals
  \newcommand{\E}{\mathbb{E}}    % expectation
  \renewcommand{\P}{\mathbb{P}}  % probability
  \DeclareMathOperator{\logit}{logit}
  \DeclareMathOperator{\logistic}{logistic}
  \DeclareMathOperator{\sd}{sd}
  \DeclareMathOperator{\var}{var}
  \DeclareMathOperator{\cov}{cov}
  \DeclareMathOperator{\cor}{cor}
  \DeclareMathOperator{\Normal}{Normal}
  \DeclareMathOperator{\LogNormal}{logNormal}
  \DeclareMathOperator{\Poisson}{Poisson}
  \DeclareMathOperator{\Beta}{Beta}
  \DeclareMathOperator{\Binom}{Binomial}
  \DeclareMathOperator{\Gam}{Gamma}
  \DeclareMathOperator{\Exp}{Exponential}
  \DeclareMathOperator{\Cauchy}{Cauchy}
  \DeclareMathOperator{\Unif}{Unif}
  \DeclareMathOperator{\Dirichlet}{Dirichlet}
  \DeclareMathOperator{\Wishart}{Wishart}
  \DeclareMathOperator{\StudentsT}{StudentsT}
  \DeclareMathOperator{\Weibull}{Weibull}

  \newcommand{\given}{\;\vert\;}
  \]
  </div>
</head>
<body>
<h1 id="homework-number-14-drosophila-viruses">Homework number 14: Drosophila viruses</h1>
<p><strong>Assignment:</strong> Your task is to use Rmarkdown to write a short report, readable by a technically literate person. The code you used should not be visible in the final report (unless you have a good reason to show it).</p>
<p><strong>Due:</strong> Submit your work via Canvas by the end of the day (midnight) on Friday, February 11th. Please submit <strong>both the Rmd file and the resulting html or pdf file.</strong> You can work with other members of class, but I expect each of you to construct and run all of the scripts yourself.</p>
<h2 id="the-problem">The Problem</h2>
<p>In this project you will analyze data from the paper <em><a href="https://elifesciences.org/articles/46440">Host-pathogen coevolution increases genetic variation in susceptibility to infection</a></em> (Duxbury et al 2019, eLife), which studies whether species (of <em>Drosophila</em> fruit fly) have more genetic variation for resistance to viruses that naturally infect that species than to viruses from other species. You will analyze their data from <em>Drosophila melanogaster</em>: the experimenters kept 230 inbred lines (called “families” in the data) of fruit fly, and measured viral loads in individual flies after infecting them with one of three types of sigma virus: one found in <em>Drosophila melanogaster</em> (“M” in the data), one found in <em>Drosophila obscura</em> (“O” in the data), and one found in <em>Drosophila affinis</em> (“A” in the data).</p>
<p>The data are found in <a href="Dmel_genotype_data.csv">Dmel_genotype_data.csv</a>, and columns in the data file are:</p>
<ul>
<li><code>family</code>: the ID of inbred line of flies</li>
<li><code>day</code>: which day of the experiment the measurements were performed on (included as a potential confounding factor)</li>
<li><code>chkov</code>: whether the family carries viral resistance alleles at the CHKov1 locus</li>
<li><code>ref2p</code>: whether the family carries viral resistance alleles at the p62 (Ref(2)p) locus</li>
<li><code>O</code>: log10 viral load in response to infection with the <em>Drosophila obscura</em> sigma virus</li>
<li><code>A</code>: log10 viral load in response to infection with the <em>Drosophila affinis</em> sigma virus</li>
<li><code>M</code>: log10 viral load in response to infection with the <em>Drosophila melanogaster</em> sigma virus</li>
</ul>
<p>You should describe the data, and then fit a multivariate mixed-effects model using the following brms code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="kw">brm</span>(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a>    <span class="kw">bf</span>(<span class="kw">mvbind</span>(A, M, O) <span class="op">|</span><span class="st"> </span><span class="kw">mi</span>() <span class="op">~</span><span class="st"> </span>chkov <span class="op">+</span><span class="st"> </span>ref2p <span class="op">+</span><span class="st"> </span>(<span class="dv">1</span><span class="op">|</span>n<span class="op">|</span>family) <span class="op">+</span><span class="st"> </span>(<span class="dv">1</span><span class="op">|</span>day))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a>    <span class="op">+</span><span class="st"> </span><span class="kw">set_rescor</span>(<span class="ot">FALSE</span>),</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>    <span class="dt">data=</span>geno,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a>    <span class="dt">chains=</span><span class="dv">3</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>)</span></code></pre></div>
<p>(Notice that there’s a fair amount of missing data, as not every virus was measured on every day. The <code>| mi()</code> term asks brms to <em>impute</em> these missing values, effectively treating the missing values as parameters to be estimated.)</p>
<p>Interpret the results to answer the following questions (including appropriate measures of uncertainty):</p>
<ol type="1">
<li><p>How much do resistance loci at the CHKov1 and Ref(2)p loci affect viral loads for the three viruses? The experimenter’s hypothesis is that they would affect the coevolved “M” virus more than the others. (<em>Note:</em> here <code>conditional_effects()</code> will lead you astray; instead, use the posterior distribution of the corresponding parameter in the model.)</p></li>
<li><p>How much do different families differ in their average viral load after infection with each of the three viruses, after accounting for the resistance loci, and is this family-specific response correlated between the three viruses? (This measures remaining genetic variation for resistance, and asks whether the genetic variation for resistance to one virus is effective against the others.)</p></li>
<li><p>For future experiments, what would we predict would be the range of viral loads for flies in the A13_17 family in response infection with the “O” virus? (<em>Hint: use <code>posterior_predict( )</code>; this is possible because of the missing data imputation with <code>|mi()</code>.</em>)</p></li>
</ol>
</body>
</html>
