---
title: "Poisson linear models"
author: "Peter Ralph"
date: "1 December 2020 -- Advanced Biological Statistics"
---

```{r setup, include=FALSE}
fig.dim <- 5
knitr::opts_chunk$set(fig.width=2*fig.dim,
                      fig.height=fig.dim,
                      fig.align='center')
set.seed(23)
library(tidyverse)
library(rstan)
library(matrixStats)
options(mc.cores = parallel::detectCores())
```


# Count data

## A hypothetical situation:

1. We have **counts** of transcript numbers,

2. from some individuals of different **ages**
   and past **exposures** to solar irradiation,

3. of two **genotypes**.

. . .

*Model:*

* Counts are **Poisson**,

* with mean that depends on age and exposure,

* but effect of exposure depends on genotype.



-------------------------

::: {.columns}
::::::::: {.column width="50%"}


1. Counts are **Poisson**,

2. with mean that depends on age and exposure,

3. but effect of exposure depends on genotype.

:::
:::::::::: {.column width="50%"}


$$\begin{aligned}
    Z_i &\sim \Poisson(\mu_i) \\
\end{aligned}$$

:::
:::::::::::



-------------------------

::: {.columns}
::::::::: {.column width="50%"}


1. Counts are **Poisson**,

2. with mean that depends on age and exposure,

3. but effect of exposure depends on genotype.

:::
:::::::::: {.column width="50%"}


$$\begin{aligned}
    Z_i &\sim \Poisson(\mu_i) \\
    \mu_i &= a + b \times \text{age}_i + c \times \text{exposure}_i 
\end{aligned}$$

:::
:::::::::::



-------------------------

::: {.columns}
::::::::: {.column width="50%"}


1. Counts are **Poisson**,

2. with mean that depends on age and exposure,

3. but effect of exposure depends on genotype.

:::
:::::::::: {.column width="50%"}


$$\begin{aligned}
    Z_i &\sim \Poisson(\mu_i) \\
    \mu_i &= \exp\left( a_{g_i} + b \times \text{age}_i + c_{g_i} \times \text{exposure}_i \right)
\end{aligned}$$

:::
:::::::::::

# Poisson modeling, in practice

## The data


```{r sim_counts, include=FALSE, cache=TRUE}
true_params <- list(a=c(0, 2),
                    b=1/20,
                    c=c(1/10, -1/10),
                    sigma=0.7)
nsamples <- 500
data <- data.frame(genotype=sample(c(1,2), nsamples, 
                                   replace=TRUE),
                   age = rgamma(nsamples, 10, 1),
                   exposure = rexp(nsamples, 0.2))
data$y <- with(data, true_params$a[genotype] +
                      true_params$b * age +
                      true_params$c[genotype] * exposure)
data$mu <- exp(rnorm(nrow(data), mean=data$y, 
                     sd=true_params$sigma))
data$counts <- rpois(nsamples, data$mu)
```

```{r save_data, include=FALSE}
write.table(data, file="data/poisson_counts_data.tsv", row.names=FALSE)
write.table(data[,c(1,2,3,6)], file="data/poisson_counts.tsv", row.names=FALSE)
```

```{r plot_counts, echo=FALSE, fig.width=3*fig.dim, fig.height=1.5*fig.dim}
layout(t(1:2))
plot(counts ~ age, col=genotype, data=data, pch=20)
for (g in seq_along(unique(data$genotype))) {
    # abline(coef(lm(counts ~ age, data=subset(data, genotype == g))), col=g)
    with(subset(data, genotype == g), lines(lowess(age, counts), col=g))
}
plot(counts ~ exposure, col=genotype, data=data, pch=20)
for (g in seq_along(unique(data$genotype))) {
    # abline(coef(lm(counts ~ exposure, data=subset(data, genotype == g))), col=g)
    with(subset(data, genotype == g), lines(lowess(exposure, counts), col=g))
}
legend("topright", pch=20, col=1:2,
       legend=paste("genotype", 1:2))
```

## Your turn: use `glm( )`


$$\begin{aligned}
    Z_i &\sim \Poisson(\mu_i) \\
    \mu_i &= \exp\left( a_{g_i} + b \times \text{age}_i + c_{g_i} \times \text{exposure}_i \right)
\end{aligned}$$


Here are [the data](data/poisson_counts.tsv):
```{r show_data}
countdata <- read.table("data/poisson_counts.tsv", header=TRUE)
head(countdata)
```

## Write a Stan block

::: {.columns}
::::::::: {.column width="50%"}


1. Counts are **Poisson**,

2. with mean that depends on age and exposure,

3. but effect of exposure depends on genotype.

$$\begin{aligned}
    Z_i &\sim \Poisson(y_i) \\
    y_i &= \exp(a_{g_i} + b \times \text{age}_i \\
        &\qquad {} + c_{g_i} \times \text{exposure}_i )
\end{aligned}$$

:::
:::::::::: {.column width="50%"}

<!-- 
    NOTE: do this online, asking class what needs to be included
    and walking through how to include it 
-->

```
data {
    int N;
    int counts[N];
    vector[N] age;
    vector[N] exposure;
    int genotype[N];
    int ngeno;
}
parameters {
    real a[ngeno]; // intercepts
    real b; // slope for age
    real c[ngeno]; // slopes for exposure
}
model {
    vector[N] y; // means
    y = exp(a[genotype] + b .* age 
            + c[genotype] .* exposure);
    counts ~ poisson(y);
    // implicitly flat priors
}
```

:::
:::::::::::


## The result

*Note:* scaling the data helps Stan find the right scale to move on.

```{r run_simple_pois, cache=TRUE, dependson=c("sim_counts", "simple_poisson")}
fit1 <- stan(file="data/simple_poisson.stan",
             data=with(data, list(N=length(counts),
                               counts=counts,
                               age=age,
                               expo=exposure,
                               geno=genotype)),
             control=list(max_treedepth=12),
             iter=1000, chains=3)
```

-----------

```{r printit}
post1 <- extract(fit1)
print(fit1)
```

## 

```{r not_warmup}
stan_trace(fit1, pars=c("a","b","c","lp__"), inc_warmup=FALSE)
```

## 

Posterior distributions of the parameters:
```{r true_fit_1, echo=FALSE}
layout(matrix(1:6, nrow=2, byrow=TRUE))
with(data, {
    hist(post1$a[,1], breaks=50, main='a[1]')
    #     xlim=range(post1$a[,1], true_params$a[1]))
    # abline(v=true_params$a[1], col='red', lwd=2)
    hist(post1$a[,2], breaks=50, main='a[2]')
    #     xlim=range(post1$a[,2], true_params$a[2]))
    # abline(v=true_params$a[2], col='red', lwd=2)
    hist(post1$b/sd(age), breaks=50, main='b')
    #     xlim=range(true_params$b, post1$b/sd(age)))
    # abline(v=true_params$b, col='red', lwd=2)
    hist(post1$c[,1]/sd(exposure), breaks=50, main='c[1]')
    #     xlim=range(post1$c[,1]/sd(exposure), true_params$c[1]))
    # abline(v=true_params$c[1], col='red', lwd=2)
    hist(post1$c[,2]/sd(exposure), breaks=50, main='c[2]')
    #     xlim=range(post1$c[,2]/sd(exposure), true_params$c[2]))
    # abline(v=true_params$c[2], col='red', lwd=2)
 })
```


# Goodness of fit

## Posterior predictive simulations

Let's simulate up data *under this model* to check for goodness of fit.

. . .

One hundred times, we'll

1. pick a set of parameters from the posterior
2. compute the vector of means (`mu`)
3. simulate a vector of Poisson counts with mean `mu`


## 


::: {.columns}
:::::::: {.column width="60%"}

```{r post_sims1}
do_sim <- function () {
    k <- sample.int(nrow(post1$a), 1)
    params1 <- list(a=post1$a[k,],
                    b=post1$b[k],
                    c=post1$c[k,])
    mu1 <- with(data,
                    exp(params1$a[genotype] 
                        + params1$b * age
                        + params1$c[genotype] * exposure))
    rpois(length(mu1), mu1)
}
# 100 datasets:
sim1 <- replicate(100, do_sim())
```

:::
:::::::::: {.column width="40%"}

```
model {
    vector[N] mu;
    mu = exp(a[geno] 
             + b * age 
             + c[geno] 
               .* expo);
    counts ~ poisson(mu);
```

:::
:::::::::::

## True data are *overdispersed* relative to our simulations

```{r plot_post_sims1, echo=FALSE, fig.height=1.5*fig.dim}
post_mean_mu <- with(list2env(data),
                     exp(colMeans(post1$a)[genotype]
                         + mean(post1$b) * age
                         + colMeans(post1$c)[genotype] * exposure))
plot(data$counts[order(post_mean_mu)], ylab="counts", ylim=range(sim1), type='n')
segments(x0=seq_len(nrow(data)),
         y0=rowMins(sim1)[order(post_mean_mu)],
         y1=rowMaxs(sim1)[order(post_mean_mu)])
points(data$counts[order(post_mean_mu)], pch=20, col='red')
legend("topleft", pch=c(20,NA), lty=c(NA,1), legend=c("observed", "simulated range"), col=c('red', 'black'))
```

