<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang xml:lang>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>HW02</title>
  <style>
html {
line-height: 1.5;
font-family: Georgia, serif;
font-size: 20px;
color: #1a1a1a;
background-color: #fdfdfd;
}
body {
margin: 0 auto;
max-width: 36em;
padding-left: 50px;
padding-right: 50px;
padding-top: 50px;
padding-bottom: 50px;
hyphens: auto;
overflow-wrap: break-word;
text-rendering: optimizeLegibility;
font-kerning: normal;
}
@media (max-width: 600px) {
body {
font-size: 0.9em;
padding: 1em;
}
h1 {
font-size: 1.8em;
}
}
@media print {
body {
background-color: transparent;
color: black;
font-size: 12pt;
}
p, h2, h3 {
orphans: 3;
widows: 3;
}
h2, h3, h4 {
page-break-after: avoid;
}
}
p {
margin: 1em 0;
}
a {
color: #1a1a1a;
}
a:visited {
color: #1a1a1a;
}
img {
max-width: 100%;
}
h1, h2, h3, h4, h5, h6 {
margin-top: 1.4em;
}
h5, h6 {
font-size: 1em;
font-style: italic;
}
h6 {
font-weight: normal;
}
ol, ul {
padding-left: 1.7em;
margin-top: 1em;
}
li > ol, li > ul {
margin-top: 0;
}
blockquote {
margin: 1em 0 1em 1.7em;
padding-left: 1em;
border-left: 2px solid #e6e6e6;
color: #606060;
}
code {
font-family: Menlo, Monaco, 'Lucida Console', Consolas, monospace;
font-size: 85%;
margin: 0;
}
pre {
margin: 1em 0;
overflow: auto;
}
pre code {
padding: 0;
overflow: visible;
overflow-wrap: normal;
}
.sourceCode {
background-color: transparent;
overflow: visible;
}
hr {
background-color: #1a1a1a;
border: none;
height: 1px;
margin: 1em 0;
}
table {
margin: 1em 0;
border-collapse: collapse;
width: 100%;
overflow-x: auto;
display: block;
font-variant-numeric: lining-nums tabular-nums;
}
table caption {
margin-bottom: 0.75em;
}
tbody {
margin-top: 0.5em;
border-top: 1px solid #1a1a1a;
border-bottom: 1px solid #1a1a1a;
}
th {
border-top: 1px solid #1a1a1a;
padding: 0.25em 0.5em 0.25em 0.5em;
}
td {
padding: 0.125em 0.5em 0.25em 0.5em;
}
header {
margin-bottom: 4em;
text-align: center;
}
#TOC li {
list-style: none;
}
#TOC ul {
padding-left: 1.3em;
}
#TOC > ul {
padding-left: 0;
}
#TOC a:not(:hover) {
text-decoration: none;
}
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
width: 0.8em;
margin: 0 0.8em 0.2em -1.6em;
vertical-align: middle;
}
</style>
  <div style="display: none">
  \[
  %%
  % Add your macros here; they&#39;ll be included in pdf and html output.
  %%

  \newcommand{\R}{\mathbb{R}}    % reals
  \newcommand{\E}{\mathbb{E}}    % expectation
  \renewcommand{\P}{\mathbb{P}}  % probability
  \DeclareMathOperator{\logit}{logit}
  \DeclareMathOperator{\logistic}{logistic}
  \DeclareMathOperator{\sd}{sd}
  \DeclareMathOperator{\var}{var}
  \DeclareMathOperator{\cov}{cov}
  \DeclareMathOperator{\cor}{cor}
  \DeclareMathOperator{\Normal}{Normal}
  \DeclareMathOperator{\LogNormal}{logNormal}
  \DeclareMathOperator{\Poisson}{Poisson}
  \DeclareMathOperator{\Beta}{Beta}
  \DeclareMathOperator{\Binom}{Binomial}
  \DeclareMathOperator{\Gam}{Gamma}
  \DeclareMathOperator{\Exp}{Exponential}
  \DeclareMathOperator{\Cauchy}{Cauchy}
  \DeclareMathOperator{\Unif}{Unif}
  \DeclareMathOperator{\Dirichlet}{Dirichlet}
  \DeclareMathOperator{\Wishart}{Wishart}
  \DeclareMathOperator{\StudentsT}{StudentsT}
  \DeclareMathOperator{\Weibull}{Weibull}

  \newcommand{\given}{\;\vert\;}
  \]
  </div>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h1 id="homework-2-testing-for-heterogeneity">Homework 2: Testing for
heterogeneity</h1>
<p><strong>Assignment:</strong> Your task is to use Rmarkdown to write a
short report, readable by a technically literate person. The code you
used should not be visible in the final report (unless you have a good
reason to show it).</p>
<p><strong>Due:</strong> Submit your work via Canvas by the end of the
day (midnight) on Thursday, October 24th. Please submit <strong>both the
Rmd file <em>and</em> the resulting html or pdf file</strong>. You can
work with other members of class, but I expect each of you to construct
and run all of the scripts yourself.</p>
<h2 id="the-problem">The problem</h2>
<p>You are working with a group studying immune response to different
types of seasonal flu. From a large number of blood samples from people
with the flu but otherwise healthy, the group has determined for each
person (a) the strain of flu, (b) the strength of the immune response
(in units of pg/mL of interleukin-1β), and (c) the person’s genotype at
an immune-regulating gene.</p>
<p>The group is interested in a fine-scale classification of flu, so
they have divided the flu samples up into a fairly large number of
distinct strains. The main question they are interested in is whether
the mean strength of immune response (measured using interleukin-1β
concentration) differs by flu strains; next they are interested in which
strains (if any) tend to induce stronger, or weaker, responses. Of
course, they would like to know how any differences compare to natural
variation in immune response. In previous studies, mean immune response
differed by genotype of the immune-regulating gene, so the analysis
should take this into account. You should <em>(a)</em> look at the data,
then answer these questions with a permutation test. More detailed
instructions are below, but your report should explain, in your own
words, how each analysis works what the conclusions are, and what to
conclude after looking at both sets of results.</p>
<p>The data are available <a href="flu.tsv">here</a>. <em>Note:</em>
this is fake data; for real information on flu strains see <a href="https://nextflu.org">nextflu.org</a>.</p>
<p><strong>(a)</strong> Make an <em>interaction plot</em> for this
question: the x-axis should show the various flu strains; the y-axis
should show immune response, and there should be one line for each
genotype, connecting the mean response to each flu strain for people of
that genotype. The plot should show not just the means, but also the
range of variation in the data.</p>
<p><strong>(b)</strong> Use as the test statistic for your permutation
test the <em>standard deviation of strain means, averaged across
genotypes</em>. In other words, compute, separately by genotype, the
standard deviation of per-strain mean responses, then average the
resulting values you get across genotypes. Since we want to look at
variation between strains after controlling for genotype, you should
shuffle immune responses <em>separately</em> for each genotype. Note
that this standard deviation quantifies variation in mean immune
response between strains by genotype; the permutation test gives a
measure of significance.</p>
</body>
</html>
