<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph">
  <title>Clustering and dimension reduction</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="reveal.js/css/reveal.css">
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="reveal.js/css/theme/simple.css" id="theme">
  <!-- Printing and PDF exports -->
  <script>
    var link = document.createElement( 'link' );
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = window.location.search.match( /print-pdf/gi ) ? 'reveal.js/css/print/pdf.css' : 'reveal.js/css/print/paper.css';
    document.getElementsByTagName( 'head' )[0].appendChild( link );
  </script>
  <!--[if lt IE 9]>
  <script src="reveal.js/lib/js/html5shiv.js"></script>
  <![endif]-->
  <style type="text/css">
  
  .reveal { font-size: 30px; }
  
  .reveal h1 { font-size: 1.5em; } 
  
  .reveal h2 { font-size: 1.2em; } 
  
  .reveal .slides { text-align: left; }
  
  .reveal .slides figure { text-align: center; }
  
  .reveal figcaption { display: none; }
  
  </style>
  <div style="display: none">
  \[
  %%
  % Add your macros here; they'll be included in pdf and html output.
  %%
  
  \newcommand{\R}{\mathbb{R}}    % reals
  \newcommand{\E}{\mathbb{E}}    % expectation
  \renewcommand{\P}{\mathbb{P}}  % probability
  \DeclareMathOperator{\logit}{logit}
  \DeclareMathOperator{\logistic}{logistic}
  \DeclareMathOperator{\sd}{sd}
  \DeclareMathOperator{\var}{var}
  \DeclareMathOperator{\cov}{cov}
  \DeclareMathOperator{\Normal}{Normal}
  \DeclareMathOperator{\Poisson}{Poisson}
  \DeclareMathOperator{\Beta}{Beta}
  \DeclareMathOperator{\Binom}{Binomial}
  \DeclareMathOperator{\Gam}{Gamma}
  \DeclareMathOperator{\Exp}{Exponential}
  \DeclareMathOperator{\Cauchy}{Cauchy}
  \DeclareMathOperator{\Unif}{Unif}
  \DeclareMathOperator{\Dirichlet}{Dirichlet}
  \DeclareMathOperator{\Wishart}{Wishart}
  
  \newcommand{\given}{\;\vert\;}
  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Clustering and dimension reduction</h1>
  <p class="author">Peter Ralph</p>
  <p class="date">25 February 2019 – Advanced Biological Statistics</p>
</section>

<section><section id="visualizing-expression-space" class="title-slide slide level1"><h1>Visualizing expression space</h1></section><section id="a-conceptual-model" class="slide level2">
<h2>A conceptual model</h2>
<p>Let’s build a <em>conceptual</em> model for descriptive analysis of “mixture” expression data.</p>
<div class="fragment">
<p><strong>Data:</strong> expression data from tissue samples that consist of various <em>mixtures</em> of different cell types.</p>
</div>
<div class="fragment">
<p><strong>Goal:</strong> identify shared coexpression patterns corresponding to <em>cell type</em>.</p>
</div>
<div class="fragment">
<p><em>Similar situations:</em> identify different developmental stages from whole-organism expression; common community structures from metagenomic data.</p>
</div>
</section><section class="slide level2">

<ol type="1">
<li><p>Each cell type has a typical set of <em>mean</em> expression levels.</p></li>
<li><p>Each sample is composed of a mixture of cell types, defined by the proportions that come from each type.</p></li>
</ol>
</section><section class="slide level2">

<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Mean expression by cell type.</p></li>
<li><p>Cell type proportions by sample.</p></li>
</ol>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p><span class="math inline">\(x_{kj}\)</span> : Mean expression of gene <span class="math inline">\(j\)</span> in cell type <span class="math inline">\(k\)</span>.</p></li>
<li><p><span class="math inline">\(w_{ik}\)</span> : Proportion of sample <span class="math inline">\(i\)</span> of cell type <span class="math inline">\(k\)</span>.</p></li>
</ol>
<p><span class="math inline">\(Z_{ij}\)</span> : expression level in sample <span class="math inline">\(i\)</span> of gene <span class="math inline">\(j\)</span>.</p>
<p><span class="math display">\[\begin{aligned}
        Z_{ij} \approx \sum_{k=1}^K w_{ik} x_{kj} .
   \end{aligned}\]</span></p>
</div>
</div>
</section></section>
<section><section id="nonnegative-matrix-factorization" class="title-slide slide level1"><h1>Nonnegative matrix factorization</h1></section><section id="aka-nmf" class="slide level2">
<h2>… aka “NMF”</h2>
<p>We are <em>decomposing</em> <span class="math inline">\(Z\)</span> into the product of two lower-dimensional, nonnegative factors:</p>
<p><span class="math display">\[\begin{aligned}
    Z_{ij} &amp;\approx \sum_k w_{ik} x_{kj} \\
    w_{ik} &amp;\ge 0 \\
    x_{kj} &amp;\ge 0 .
\end{aligned}\]</span></p>
</section><section id="a-simple-nmf-model" class="slide level2">
<h2>A simple NMF model</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">simple_nmf &lt;-<span class="st"> </span><span class="kw">stan_model</span>(<span class="dt">model_code=</span><span class="st">&quot;</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="st">data {</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="st">    int N; // samples</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="st">    int L; // variables</span></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="st">    int K; // factors</span></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="st">    real Z[L,N];</span></a>
<a class="sourceLine" id="cb1-7" title="7"><span class="st">}</span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="st">parameters {</span></a>
<a class="sourceLine" id="cb1-9" title="9"><span class="st">    matrix&lt;lower=0&gt;[L,K] x;</span></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="st">    matrix&lt;lower=0&gt;[K,N] w;</span></a>
<a class="sourceLine" id="cb1-11" title="11"><span class="st">    real&lt;lower=0&gt; sigma;</span></a>
<a class="sourceLine" id="cb1-12" title="12"><span class="st">}</span></a>
<a class="sourceLine" id="cb1-13" title="13"><span class="st">model {</span></a>
<a class="sourceLine" id="cb1-14" title="14"><span class="st">    for (j in 1:L) {</span></a>
<a class="sourceLine" id="cb1-15" title="15"><span class="st">        Z[j] ~ normal(x[j] * w, sigma);</span></a>
<a class="sourceLine" id="cb1-16" title="16"><span class="st">    }</span></a>
<a class="sourceLine" id="cb1-17" title="17"><span class="st">}</span></a>
<a class="sourceLine" id="cb1-18" title="18"><span class="st">&quot;</span>)</a></code></pre></div>
</section><section id="relationship-to-pca" class="slide level2">
<h2>Relationship to PCA</h2>
<p>PCA finds <span class="math inline">\(w\)</span> and <span class="math inline">\(z\)</span> to minimize <span class="math display">\[\begin{aligned}
    \sum_{ij} \| Z_{ij} - \sum_k w_{ik} x_{kj} \|^2 .
\end{aligned}\]</span></p>
<p>In other words, it is the maximum-likelihood solution to <span class="math display">\[\begin{aligned}
    Z_{ij} &amp;\sim \Normal(\sum_k w_{ik} x_{kj}, \sigma^2) .
\end{aligned}\]</span> (The eigenvectors are the columns of <span class="math inline">\(x\)</span>, and the eigenvectors are related to the size of <span class="math inline">\(w\)</span> and <span class="math inline">\(x\)</span>.)</p>
</section><section id="pca-in-stan" class="slide level2">
<h2>PCA, in Stan</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">stan_pca &lt;-<span class="st"> </span><span class="kw">stan_model</span>(<span class="dt">model_code=</span><span class="st">&quot;</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="st">data {</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="st">    int N; // samples</span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="st">    int L; // variables</span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="st">    int K; // factors</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="st">    real Z[L,N];</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="st">}</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="st">parameters {</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="st">    matrix[L,K] x;</span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="st">    matrix[K,N] w;</span></a>
<a class="sourceLine" id="cb2-11" title="11"><span class="st">    real&lt;lower=0&gt; sigma;</span></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="st">}</span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="st">model {</span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="st">    for (j in 1:L) {</span></a>
<a class="sourceLine" id="cb2-15" title="15"><span class="st">        Z[j] ~ normal(x[j] * w, sigma);</span></a>
<a class="sourceLine" id="cb2-16" title="16"><span class="st">    }</span></a>
<a class="sourceLine" id="cb2-17" title="17"><span class="st">}</span></a>
<a class="sourceLine" id="cb2-18" title="18"><span class="st">&quot;</span>)</a></code></pre></div>
<p><em>(note: needs some priors to work well; see <a href="https://arxiv.org/abs/1603.00788">here</a>.)</em></p>
<!--
## PCA in Stan


```r
nvars <- 20
nsamp <- 200
xy <- matrix(rnorm(2*nvars), nrow=2)
z <- matrix(rnorm(2*nsamp), ncol=2) %*% xy
stpca <- optimizing(stan_pca, data=list(N=nsamp, L=nvars, K=2, Z=t(z)))
stan_x <- t(matrix(stpca$par[grepl("^w",names(stpca$par))], nrow=2))
```

----------------


```r
usual_pca <- eigen(cov(t(z)))
layout(t(1:2))
plot(usual_pca$vectors[,1:2], cex=2,
     col=rainbow(10)[as.numeric(cut(xy[1,], 10))], 
     pch=as.numeric(cut(xy[2,], 6)),
     xlab="PC 1", ylab="PC 2")
plot(stan_x, cex=2,
     col=rainbow(10)[as.numeric(cut(xy[1,], 10))], 
     pch=as.numeric(cut(xy[2,], 6)),
     xlab="Stan PC 1", ylab="Stan PC 2")
```

<img src="figure/Week_18_Lecture/plot_pcas-1.png" title="plot of chunk plot_pcas" alt="plot of chunk plot_pcas" style="display: block; margin: auto;" />
-->
</section></section>
<section><section id="stochastic-minute" class="title-slide slide level1"><h1>Stochastic minute</h1></section><section id="the-dirichlet-distribution" class="slide level2">
<h2>the Dirichlet distribution</h2>
<p>A random set of <span class="math inline">\(k\)</span> <em>proportions</em> <span class="math inline">\(0 \le P_i \le 1\)</span> has a <span class="math inline">\(\Dirichlet(\alpha_1, \ldots, \alpha_k)\)</span> if it has probability density <span class="math display">\[\begin{aligned}
    \frac{1}{B(\alpha)} \prod_{i=1}^k p_i^{\alpha_i} 
\end{aligned}\]</span> over the set of possible values <span class="math display">\[\begin{aligned}
    P_1 + \cdots + P_k = 1 .
\end{aligned}\]</span></p>
<div class="fragment">
<ol type="1">
<li><p>This is useful as a prior on <em>proportions</em>.</p></li>
<li><p>The <em>mean</em> is <span class="math display">\[ \left( \frac{\alpha_1}{\sum_j \alpha_j}, \frac{\alpha_2}{\sum_j \alpha_j}, \cdots, \frac{\alpha_k}{\sum_j \alpha_j} \right) . \]</span></p></li>
<li><p>This generalized the Beta: if <span class="math inline">\(X \sim \Beta(a, b)\)</span> then <span class="math inline">\((X, 1-X) \sim \Dirichlet(a, b)\)</span>.</p></li>
<li><p>Marginal distributions are Beta distributed: <span class="math inline">\(P_i \sim \Beta(\alpha_i, \sum_{j=1}^k \alpha_j - \alpha_i)\)</span>.</p></li>
</ol>
</div>
</section><section class="slide level2">

<ol start="5" type="1">
<li>If <span class="math inline">\(X_i \sim \Gamma(\text{shape}=\alpha_i)\)</span>, and <span class="math display">\[\begin{aligned}
 P_i = X_i / \sum_{j=1}^k X_j
\end{aligned}\]</span> then <span class="math inline">\(P \sim \Dirichlet(\alpha)\)</span>.</li>
</ol>
</section><section id="simplex-parameters" class="slide level2">
<h2>“Simplex” parameters</h2>
<p>“The <span class="math inline">\(k\)</span>-simplex” is the set of <em>proportions</em>, i.e., nonnegative numbers <span class="math inline">\(p\)</span> satisfying <span class="math display">\[\begin{aligned}
    p_1 + \cdots p_k = 1 .
\end{aligned}\]</span></p>
<pre><code>parameters {
    simplex[K] p;
}
model {
    p ~ dirichlet(alpha);
}</code></pre>
</section></section>
<section><section id="back-to-expression-space" class="title-slide slide level1"><h1>Back to expression space</h1></section><section class="slide level2">

<ol type="1">
<li><p>Each cell type has a typical set of <em>mean</em> expression levels.</p></li>
<li><p>Each sample is composed of a mixture of cell types, defined by the proportions that come from each type.</p></li>
<li><p>Mean expression levels differ between cell types for only some of the genes.</p></li>
<li><p>Some samples are <em>noisier</em> than others.</p></li>
</ol>
</section><section class="slide level2">

<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Mean expression by cell type.</p></li>
<li><p>Cell type proportions by sample.</p></li>
</ol>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p><span class="math inline">\(x_{kj}\)</span> : Mean expression of gene <span class="math inline">\(j\)</span> in cell type <span class="math inline">\(k\)</span>.</p></li>
<li><p><span class="math inline">\(w_{ik}\)</span> : Proportion of sample <span class="math inline">\(i\)</span> of cell type <span class="math inline">\(k\)</span>.</p></li>
</ol>
<p><span class="math inline">\(Z_{ij}\)</span> : expression in sample <span class="math inline">\(i\)</span> of gene <span class="math inline">\(j\)</span>.</p>
<p><span class="math display">\[\begin{aligned}
        Z_{ij} \approx \sum_{k=1}^K w_{ik} x_{kj} .
   \end{aligned}\]</span></p>
</div>
</div>
</section><section class="slide level2">

<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Mean expression by cell type.</p></li>
<li><p>Cell type proportions by sample.</p></li>
<li><p>Mean expression levels differ between cell types for only some of the genes.</p></li>
<li><p>Some samples are <em>noisier</em> than others.</p></li>
</ol>
</div><div class="column" style="width:50%;">
<p><span class="math inline">\(Z_{ij}\)</span> : expression level in sample <span class="math inline">\(i\)</span> of gene <span class="math inline">\(j\)</span>.</p>
<p><span class="math display">\[\begin{aligned}
        Z_{ij} \approx \sum_{k=1}^K w_{ik} x_{kj} .
   \end{aligned}\]</span></p>
<ol start="3" type="1">
<li><p><span class="math inline">\(y_j\)</span>, <span class="math inline">\(\eta_j\)</span> : mean and SD of expression of gene <span class="math inline">\(j\)</span> across <em>all</em> cell types; shrink <span class="math inline">\(x_{kj}\)</span> towards <span class="math inline">\(y_j\)</span>.</p></li>
<li><p><em>(omit this)</em></p></li>
</ol>
</div>
</div>
</section><section class="slide level2">

<div class="columns">
<div class="column" style="width:50%;">
<pre><code>data {
  int N; // # samples
  int L; // # genes
  int K; // # cell types
  int Z[N,L];
}</code></pre>
<pre><code>parameters {
  matrix&lt;lower=0&gt;[L,K] x;
  vector[L] y;
  simplex[K] w[N];
  vector&lt;lower=0&gt;[L] eta;
  vector&lt;lower=0&gt;[K] alpha;
  real&lt;lower=0&gt; d_alpha;
}</code></pre>
<pre><code>model {
  for (i in 1:N) {
      Z[i] ~ poisson(eta .* (x * w[i]));
      w[i] ~ dirichlet(d_alpha * alpha);
  }
  for (j in 1:K) 
      { x[,j] ~ normal(y ./ eta, 1); }
  y ~ normal(0, 20);
  alpha ~ normal(0, 1);
  d_alpha ~ exponential(0.2);
  eta ~ cauchy(0, 10);
}</code></pre>
</div><div class="column" style="width:50%;">
<ol type="1">
<li><p><span class="math inline">\(x_{kj}\)</span> : Mean expression of gene <span class="math inline">\(j\)</span> in cell type <span class="math inline">\(k\)</span>.</p></li>
<li><p><span class="math inline">\(w_{ik}\)</span> : Proportion of sample <span class="math inline">\(i\)</span> of cell type <span class="math inline">\(k\)</span>.</p></li>
</ol>
<p><span class="math display">\[\begin{aligned}
    Z_{ij} \approx \sum_k w_{ik} x_{kj} .
\end{aligned}\]</span></p>
<ol start="3" type="1">
<li><span class="math inline">\(y_j\)</span>, <span class="math inline">\(\eta_j\)</span> : mean and SD of expression of gene <span class="math inline">\(j\)</span> across <em>all</em> cell types; shrink <span class="math inline">\(x_{kj}\)</span> towards <span class="math inline">\(y_j\)</span>.</li>
</ol>
</div>
</div>
</section><section id="testing-compiles" class="slide level2">
<h2>Testing: compiles?</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">nmf_model &lt;-<span class="st"> </span><span class="kw">stan_model</span>(<span class="dt">model_code=</span>nmf_block)</a></code></pre></div>
</section><section id="testing-runs" class="slide level2">
<h2>Testing: runs?</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="kw">sampling</span>(nmf_model,</a>
<a class="sourceLine" id="cb8-2" title="2">         <span class="dt">data=</span><span class="kw">list</span>(<span class="dt">N=</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb8-3" title="3">                   <span class="dt">L=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb8-4" title="4">                   <span class="dt">K=</span><span class="dv">2</span>,</a>
<a class="sourceLine" id="cb8-5" title="5">                   <span class="dt">Z=</span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">50</span>, <span class="dv">100</span>), <span class="dt">ncol=</span><span class="dv">5</span>)),</a>
<a class="sourceLine" id="cb8-6" title="6">         <span class="dt">chains=</span><span class="dv">1</span>, <span class="dt">iter=</span><span class="dv">100</span>)</a></code></pre></div>
<pre><code>## 
## SAMPLING FOR MODEL &#39;027c9a0b92082dc735bfb6350861f7ee&#39; NOW (CHAIN 1).
## Chain 1: 
## Chain 1: Gradient evaluation took 5.7e-05 seconds
## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0.57 seconds.
## Chain 1: Adjust your expectations accordingly!
## Chain 1: 
## Chain 1: 
## Chain 1: WARNING: There aren&#39;t enough warmup iterations to fit the
## Chain 1:          three stages of adaptation as currently configured.
## Chain 1:          Reducing each adaptation stage to 15%/75%/10% of
## Chain 1:          the given number of warmup iterations:
## Chain 1:            init_buffer = 7
## Chain 1:            adapt_window = 38
## Chain 1:            term_buffer = 5
## Chain 1: 
## Chain 1: Iteration:  1 / 100 [  1%]  (Warmup)
## Chain 1: Iteration: 10 / 100 [ 10%]  (Warmup)
## Chain 1: Iteration: 20 / 100 [ 20%]  (Warmup)
## Chain 1: Iteration: 30 / 100 [ 30%]  (Warmup)
## Chain 1: Iteration: 40 / 100 [ 40%]  (Warmup)
## Chain 1: Iteration: 50 / 100 [ 50%]  (Warmup)
## Chain 1: Iteration: 51 / 100 [ 51%]  (Sampling)
## Chain 1: Iteration: 60 / 100 [ 60%]  (Sampling)
## Chain 1: Iteration: 70 / 100 [ 70%]  (Sampling)
## Chain 1: Iteration: 80 / 100 [ 80%]  (Sampling)
## Chain 1: Iteration: 90 / 100 [ 90%]  (Sampling)
## Chain 1: Iteration: 100 / 100 [100%]  (Sampling)
## Chain 1: 
## Chain 1:  Elapsed Time: 0.710808 seconds (Warm-up)
## Chain 1:                1.04427 seconds (Sampling)
## Chain 1:                1.75508 seconds (Total)
## Chain 1:</code></pre>
<pre><code>## Warning: There were 3 divergent transitions after warmup. Increasing adapt_delta above 0.8 may help. See
## http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup</code></pre>
<pre><code>## Warning: There were 31 transitions after warmup that exceeded the maximum treedepth. Increase max_treedepth above 10. See
## http://mc-stan.org/misc/warnings.html#maximum-treedepth-exceeded</code></pre>
<pre><code>## Warning: There were 1 chains where the estimated Bayesian Fraction of Missing Information was low. See
## http://mc-stan.org/misc/warnings.html#bfmi-low</code></pre>
<pre><code>## Warning: Examine the pairs() plot to diagnose sampling problems</code></pre>
<pre><code>## Inference for Stan model: 027c9a0b92082dc735bfb6350861f7ee.
## 1 chains, each with iter=100; warmup=50; thin=1; 
## post-warmup draws per chain=50, total post-warmup draws=50.
## 
##              mean se_mean     sd     2.5%      25%      50%      75%
## x[1,1]       1.19    0.12   0.84     0.18     0.62     0.95     1.52
## x[1,2]       0.93    0.14   0.79     0.11     0.45     0.65     1.19
## x[2,1]       1.39    0.47   1.10     0.11     0.54     1.07     2.11
## x[2,2]       1.05    0.12   0.77     0.04     0.47     0.90     1.28
## x[3,1]       1.49    0.29   0.98     0.13     0.58     1.36     2.10
## x[3,2]       1.16    0.15   0.81     0.13     0.53     1.12     1.52
## x[4,1]       1.33    0.10   0.72     0.16     0.91     1.33     1.62
## x[4,2]       1.05    0.25   0.89     0.00     0.31     0.86     1.59
## x[5,1]       1.94    0.33   1.23     0.33     1.07     1.54     2.60
## x[5,2]       1.62    0.17   1.03     0.08     0.93     1.37     2.54
## y[1]         9.77    6.92  16.48   -19.69    -4.64    12.59    24.19
## y[2]         9.75   13.07  25.19   -28.61    -6.47     5.02    34.16
## y[3]        24.10    9.26  21.39    -8.88     7.76    18.70    47.09
## y[4]         5.94    6.41  15.09   -21.34    -3.24     5.40    19.94
## y[5]        40.01    3.96  14.87    14.17    27.96    41.65    51.49
## w[1,1]       0.77    0.16   0.32     0.08     0.65     0.97     0.99
## w[1,2]       0.23    0.16   0.32     0.00     0.01     0.03     0.35
## w[2,1]       0.76    0.16   0.31     0.08     0.65     0.93     1.00
## w[2,2]       0.24    0.16   0.31     0.00     0.00     0.07     0.35
## w[3,1]       0.76    0.17   0.30     0.13     0.64     0.92     1.00
## w[3,2]       0.24    0.17   0.30     0.00     0.00     0.08     0.36
## w[4,1]       0.77    0.16   0.31     0.08     0.68     0.94     1.00
## w[4,2]       0.23    0.16   0.31     0.00     0.00     0.06     0.32
## w[5,1]       0.76    0.17   0.32     0.06     0.65     0.93     1.00
## w[5,2]       0.24    0.17   0.32     0.00     0.00     0.07     0.35
## w[6,1]       0.77    0.16   0.31     0.16     0.68     0.96     1.00
## w[6,2]       0.23    0.16   0.31     0.00     0.00     0.04     0.32
## w[7,1]       0.77    0.16   0.31     0.12     0.61     0.95     1.00
## w[7,2]       0.23    0.16   0.31     0.00     0.00     0.05     0.39
## w[8,1]       0.78    0.16   0.31     0.12     0.61     0.97     1.00
## w[8,2]       0.22    0.16   0.31     0.00     0.00     0.03     0.39
## w[9,1]       0.77    0.16   0.30     0.10     0.60     0.96     1.00
## w[9,2]       0.23    0.16   0.30     0.00     0.00     0.04     0.40
## w[10,1]      0.75    0.17   0.32     0.08     0.61     0.94     0.99
## w[10,2]      0.25    0.17   0.32     0.00     0.01     0.06     0.39
## eta[1]     131.11   17.50 104.77    31.83    68.49   104.77   169.52
## eta[2]     113.86   17.72 124.60    26.25    46.53    85.96   142.42
## eta[3]     105.16   23.88  91.35    32.74    47.74    80.49   128.30
## eta[4]      96.10   12.16  93.52    29.53    59.67    72.76    97.18
## eta[5]      70.69   10.19  44.27    20.59    36.89    58.75    90.14
## alpha[1]     1.09    0.10   0.73     0.20     0.54     0.88     1.61
## alpha[2]     0.53    0.35   0.73     0.02     0.05     0.18     0.72
## d_alpha      8.14    2.70   5.71     0.69     3.69     6.42    12.55
## lp__     18013.92    8.93  14.57 17994.67 17999.31 18011.47 18029.52
##             97.5% n_eff Rhat
## x[1,1]       3.15    46 0.99
## x[1,2]       2.81    32 1.04
## x[2,1]       3.59     5 1.49
## x[2,2]       2.41    39 1.02
## x[3,1]       3.14    11 1.28
## x[3,2]       2.87    28 0.99
## x[4,1]       3.22    53 0.98
## x[4,2]       2.78    12 1.17
## x[5,1]       4.95    14 0.98
## x[5,2]       3.51    36 1.04
## y[1]        30.90     6 1.06
## y[2]        45.45     4 1.89
## y[3]        60.20     5 1.31
## y[4]        30.95     6 1.00
## y[5]        66.78    14 0.99
## w[1,1]       1.00     4 1.60
## w[1,2]       0.92     4 1.60
## w[2,1]       1.00     4 1.71
## w[2,2]       0.92     4 1.71
## w[3,1]       1.00     3 1.81
## w[3,2]       0.87     3 1.81
## w[4,1]       1.00     4 1.57
## w[4,2]       0.92     4 1.57
## w[5,1]       1.00     3 1.73
## w[5,2]       0.94     3 1.73
## w[6,1]       1.00     4 1.65
## w[6,2]       0.84     4 1.65
## w[7,1]       1.00     3 1.72
## w[7,2]       0.88     3 1.72
## w[8,1]       1.00     4 1.66
## w[8,2]       0.88     4 1.66
## w[9,1]       1.00     4 1.65
## w[9,2]       0.90     4 1.65
## w[10,1]      1.00     4 1.68
## w[10,2]      0.92     4 1.68
## eta[1]     344.99    36 1.00
## eta[2]     247.98    49 1.00
## eta[3]     334.45    15 1.15
## eta[4]     332.98    59 1.02
## eta[5]     182.97    19 1.01
## alpha[1]     2.66    52 0.98
## alpha[2]     2.40     4 1.50
## d_alpha     20.43     4 1.72
## lp__     18034.18     3 2.98
## 
## Samples were drawn using NUTS(diag_e) at Tue Mar  5 22:41:23 2019.
## For each parameter, n_eff is a crude measure of effective sample size,
## and Rhat is the potential scale reduction factor on split chains (at 
## convergence, Rhat=1).</code></pre>
</section></section>
<section><section id="simulate-data" class="title-slide slide level1"><h1>Simulate data</h1></section><section id="outline" class="slide level2">
<h2>Outline</h2>
<ol type="1">
<li><p>How many cell types?</p></li>
<li><p>How many genes?</p></li>
<li><p>How many samples?</p></li>
<li><p>How much noise in expression?</p></li>
<li><p>How many genes distinguish cell types, and by how much relative to expression?</p></li>
<li><p>How many “noisy” genes? How many “noisy” samples?</p></li>
<li><p>How much variation in mixture proportions?</p></li>
</ol>
</section></section>
<section><section id="easier-case" class="title-slide slide level1"><h1>Easier case</h1></section><section class="slide level2">

<p>Sample mixture proportions (the rows of <code>w</code>), from the Dirichlet distribution with parameters $= 0.3333333, 0.3333333, 0.3333333 $.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">w &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rgamma</span>(nsamples <span class="op">*</span><span class="st"> </span>ntypes, <span class="dt">rate=</span><span class="dv">1</span>, </a>
<a class="sourceLine" id="cb15-2" title="2">                   <span class="dt">shape=</span><span class="kw">rep</span>(mean_props, <span class="dt">each=</span>nsamples)), </a>
<a class="sourceLine" id="cb15-3" title="3">            <span class="dt">ncol=</span>ntypes)</a>
<a class="sourceLine" id="cb15-4" title="4">w &lt;-<span class="st"> </span>w <span class="op">/</span><span class="st"> </span><span class="kw">rowSums</span>(w)</a>
<a class="sourceLine" id="cb15-5" title="5"><span class="kw">stopifnot</span>(<span class="kw">all</span>(<span class="kw">abs</span>(<span class="kw">rowSums</span>(w) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-8</span>))</a></code></pre></div>
</section><section class="slide level2">

<p>Sample cell type expression levels:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">mean_x &lt;-<span class="st"> </span><span class="kw">abs</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">rnorm</span>(ngenes))</a>
<a class="sourceLine" id="cb16-2" title="2">x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(mean_x, <span class="dt">each=</span>ntypes), <span class="dt">nrow=</span>ntypes)</a>
<a class="sourceLine" id="cb16-3" title="3"><span class="kw">stopifnot</span>(<span class="kw">all</span>(<span class="kw">colMeans</span>(x) <span class="op">==</span><span class="st"> </span>mean_x))</a>
<a class="sourceLine" id="cb16-4" title="4"><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>ntypes) {</a>
<a class="sourceLine" id="cb16-5" title="5">    diff_genes &lt;-<span class="st"> </span><span class="kw">sample.int</span>(ngenes, ndiff_genes)</a>
<a class="sourceLine" id="cb16-6" title="6">    x[k, diff_genes] &lt;-<span class="st"> </span><span class="kw">rgamma</span>(ndiff_genes, <span class="dt">shape=</span><span class="fl">0.5</span>, <span class="dt">scale=</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span>x[k, diff_genes]</a>
<a class="sourceLine" id="cb16-7" title="7">}</a></code></pre></div>
</section><section class="slide level2">

<p>Simulate expression:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">mean_z &lt;-<span class="st"> </span>w <span class="op">%*%</span><span class="st"> </span>x</a>
<a class="sourceLine" id="cb17-2" title="2">noisy_samples &lt;-<span class="st"> </span><span class="kw">sample.int</span>(nsamples, num_noisy_samples)</a>
<a class="sourceLine" id="cb17-3" title="3">mean_z[noisy_samples,] &lt;-<span class="st"> </span>(mean_z[noisy_samples,] </a>
<a class="sourceLine" id="cb17-4" title="4">                           <span class="op">*</span><span class="st"> </span><span class="kw">exp</span>(<span class="kw">rnorm</span>(noisy_samples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span><span class="fl">0.1</span>)))</a>
<a class="sourceLine" id="cb17-5" title="5">noisy_mean_z &lt;-<span class="st"> </span>mean_z <span class="op">*</span><span class="st"> </span><span class="kw">exp</span>(<span class="kw">rnorm</span>(nsamples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span>indiv_sd))</a>
<a class="sourceLine" id="cb17-6" title="6">Z &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(nsamples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">lambda=</span>noisy_mean_z), <span class="dt">ncol=</span>ngenes)</a></code></pre></div>
</section><section class="slide level2">

<p>Run Stan:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">ez_results &lt;-<span class="st"> </span><span class="kw">optimizing</span>(nmf_model,</a>
<a class="sourceLine" id="cb18-2" title="2">                          <span class="dt">data=</span><span class="kw">list</span>(<span class="dt">N=</span>nsamples,</a>
<a class="sourceLine" id="cb18-3" title="3">                                    <span class="dt">L=</span>ngenes,</a>
<a class="sourceLine" id="cb18-4" title="4">                                    <span class="dt">K=</span>ntypes,</a>
<a class="sourceLine" id="cb18-5" title="5">                                    <span class="dt">Z=</span>Z))</a></code></pre></div>
<p>Look at output:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">est_w &lt;-<span class="st"> </span><span class="kw">matrix</span>(ez_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^w&quot;</span>, <span class="kw">names</span>(ez_results<span class="op">$</span>par))], <span class="dt">nrow=</span>nsamples)</a>
<a class="sourceLine" id="cb19-2" title="2">est_x &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">matrix</span>(ez_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^x&quot;</span>, <span class="kw">names</span>(ez_results<span class="op">$</span>par))], <span class="dt">nrow=</span>ngenes))</a>
<a class="sourceLine" id="cb19-3" title="3">est_x &lt;-<span class="st"> </span><span class="kw">sweep</span>(est_x, <span class="dv">2</span>, ez_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^eta&quot;</span>, <span class="kw">names</span>(ez_results<span class="op">$</span>par))], <span class="st">&quot;*&quot;</span>)</a></code></pre></div>
</section><section class="slide level2">

<p>Here is the correlation between estimated <code>w</code> and observed <code>w</code>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="kw">cor</span>(est_w, w)</a></code></pre></div>
<pre><code>##            [,1]       [,2]       [,3]
## [1,] -0.4479292  0.9974603 -0.5419578
## [2,]  0.9991217 -0.4967387 -0.5112457
## [3,] -0.5412021 -0.4560523  0.9986858</code></pre>
<p>This is <strong>very</strong> good! We are estimating relative mixtures very well. (But, note they are not in the same order.)</p>
</section><section class="slide level2">

<p>Here is the correlation between estimated <code>x</code> and observed <code>x</code>:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="kw">cor</span>(<span class="kw">t</span>(est_x), <span class="kw">t</span>(x))</a></code></pre></div>
<pre><code>##           [,1]      [,2]      [,3]
## [1,] 0.6392301 0.9986823 0.4427321
## [2,] 0.9948616 0.5748984 0.5025609
## [3,] 0.5244367 0.4837158 0.9999782</code></pre>
<p>This is also very good!</p>
</section></section>
<section><section id="in-class-implementation-harder-case" class="title-slide slide level1"><h1>In-class implementation (harder case)</h1></section><section class="slide level2">

<p>Sample mixture proportions (the rows of <code>w</code>), from the Dirichlet distribution with parameters $= 0.0277778, 0.0555556, 0.0833333, 0.1111111, 0.1388889, 0.1666667, 0.1944444, 0.2222222 $.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">w &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rgamma</span>(nsamples <span class="op">*</span><span class="st"> </span>ntypes, <span class="dt">rate=</span><span class="dv">1</span>, </a>
<a class="sourceLine" id="cb24-2" title="2">                   <span class="dt">shape=</span><span class="kw">rep</span>(mean_props, <span class="dt">each=</span>nsamples)), </a>
<a class="sourceLine" id="cb24-3" title="3">            <span class="dt">ncol=</span>ntypes)</a>
<a class="sourceLine" id="cb24-4" title="4">w &lt;-<span class="st"> </span>w <span class="op">/</span><span class="st"> </span><span class="kw">rowSums</span>(w)</a>
<a class="sourceLine" id="cb24-5" title="5"><span class="kw">stopifnot</span>(<span class="kw">all</span>(<span class="kw">abs</span>(<span class="kw">rowSums</span>(w) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-8</span>))</a></code></pre></div>
</section><section class="slide level2">

<p>Sample cell type expression levels:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">mean_x &lt;-<span class="st"> </span><span class="kw">abs</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">rcauchy</span>(ngenes))</a>
<a class="sourceLine" id="cb25-2" title="2">x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(mean_x, <span class="dt">each=</span>ntypes), <span class="dt">nrow=</span>ntypes)</a>
<a class="sourceLine" id="cb25-3" title="3"><span class="kw">stopifnot</span>(<span class="kw">all</span>(<span class="kw">colMeans</span>(x) <span class="op">==</span><span class="st"> </span>mean_x))</a>
<a class="sourceLine" id="cb25-4" title="4"><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>ntypes) {</a>
<a class="sourceLine" id="cb25-5" title="5">    diff_genes &lt;-<span class="st"> </span><span class="kw">sample.int</span>(ngenes, ndiff_genes)</a>
<a class="sourceLine" id="cb25-6" title="6">    x[k, diff_genes] &lt;-<span class="st"> </span><span class="kw">rgamma</span>(ndiff_genes, <span class="dt">shape=</span><span class="fl">0.5</span>, <span class="dt">scale=</span>diff_size) <span class="op">*</span><span class="st"> </span>x[k, diff_genes]</a>
<a class="sourceLine" id="cb25-7" title="7">}</a></code></pre></div>
</section><section class="slide level2">

<p>Simulate expression:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1">mean_z &lt;-<span class="st"> </span>w <span class="op">%*%</span><span class="st"> </span>x</a>
<a class="sourceLine" id="cb26-2" title="2">noisy_samples &lt;-<span class="st"> </span><span class="kw">sample.int</span>(nsamples, num_noisy_samples)</a>
<a class="sourceLine" id="cb26-3" title="3">mean_z[noisy_samples,] &lt;-<span class="st"> </span>(mean_z[noisy_samples,] </a>
<a class="sourceLine" id="cb26-4" title="4">                           <span class="op">*</span><span class="st"> </span><span class="kw">exp</span>(<span class="kw">rnorm</span>(noisy_samples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span><span class="fl">0.1</span>)))</a>
<a class="sourceLine" id="cb26-5" title="5">noisy_mean_z &lt;-<span class="st"> </span>mean_z <span class="op">*</span><span class="st"> </span><span class="kw">exp</span>(<span class="kw">rnorm</span>(nsamples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span>indiv_sd))</a>
<a class="sourceLine" id="cb26-6" title="6">Z &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(nsamples <span class="op">*</span><span class="st"> </span>ngenes, <span class="dt">lambda=</span>noisy_mean_z), <span class="dt">ncol=</span>ngenes)</a></code></pre></div>
</section><section class="slide level2">

<p>Is there signal? It looks like there is.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1">diff8 &lt;-<span class="st"> </span>x[<span class="dv">8</span>,] <span class="op">-</span><span class="st"> </span>matrixStats<span class="op">::</span><span class="kw">colMedians</span>(x)</a>
<a class="sourceLine" id="cb27-2" title="2"><span class="kw">image</span>(<span class="kw">log</span>(<span class="kw">t</span>(Z[<span class="kw">order</span>(w[,<span class="dv">8</span>]),<span class="kw">order</span>(<span class="kw">abs</span>(diff8))])))</a></code></pre></div>
<p><img src="figure/Week_18_Lecture/image-1.png" title="plot of chunk image" alt="plot of chunk image" style="display: block; margin: auto;" /></p>
</section><section class="slide level2">

<p>Run Stan:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">opt_results &lt;-<span class="st"> </span><span class="kw">optimizing</span>(nmf_model,</a>
<a class="sourceLine" id="cb28-2" title="2">                          <span class="dt">data=</span><span class="kw">list</span>(<span class="dt">N=</span>nsamples,</a>
<a class="sourceLine" id="cb28-3" title="3">                                    <span class="dt">L=</span>ngenes,</a>
<a class="sourceLine" id="cb28-4" title="4">                                    <span class="dt">K=</span>ntypes,</a>
<a class="sourceLine" id="cb28-5" title="5">                                    <span class="dt">Z=</span>Z))</a></code></pre></div>
<p>Look at output:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1">est_w &lt;-<span class="st"> </span><span class="kw">matrix</span>(opt_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^w&quot;</span>, <span class="kw">names</span>(opt_results<span class="op">$</span>par))], <span class="dt">nrow=</span>nsamples)</a>
<a class="sourceLine" id="cb29-2" title="2">est_x &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">matrix</span>(opt_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^x&quot;</span>, <span class="kw">names</span>(opt_results<span class="op">$</span>par))], <span class="dt">nrow=</span>ngenes))</a>
<a class="sourceLine" id="cb29-3" title="3">est_x &lt;-<span class="st"> </span><span class="kw">sweep</span>(est_x, <span class="dv">2</span>, opt_results<span class="op">$</span>par[<span class="kw">grepl</span>(<span class="st">&quot;^eta&quot;</span>, <span class="kw">names</span>(opt_results<span class="op">$</span>par))], <span class="st">&quot;*&quot;</span>)</a></code></pre></div>
</section><section class="slide level2">

<p>Here is the correlation between estimated <code>w</code> and observed <code>w</code>:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="kw">cor</span>(est_w, w)</a></code></pre></div>
<pre><code>##             [,1]        [,2]        [,3]        [,4]       [,5]
## [1,] -0.06511751 -0.08548045 -0.12145422  0.99779515 -0.1207468
## [2,] -0.08882160 -0.10209420 -0.17533424 -0.15455090 -0.2230952
## [3,] -0.01964063 -0.06896749  0.99732363 -0.12371316 -0.1475214
## [4,]  0.99886737 -0.03238973 -0.02588937 -0.06523943 -0.0677335
## [5,] -0.06712511 -0.10251975 -0.14701999 -0.11864029  0.9990249
## [6,] -0.07116101 -0.11989702 -0.13666643 -0.15799797 -0.1881662
## [7,] -0.03917093  0.99906854 -0.06805167 -0.08390533 -0.1050867
## [8,] -0.10300235 -0.11957976 -0.15060561 -0.17508342 -0.2109410
##             [,6]        [,7]        [,8]
## [1,] -0.16227159 -0.15239993 -0.17445705
## [2,] -0.22241746  0.99900054 -0.24794625
## [3,] -0.13752298 -0.17502168 -0.14926717
## [4,] -0.07129504 -0.09014155 -0.09375712
## [5,] -0.18860476 -0.22231870 -0.21174459
## [6,]  0.99877567 -0.22528295 -0.23389558
## [7,] -0.11215778 -0.10425900 -0.11965410
## [8,] -0.23685561 -0.24854604  0.99801810</code></pre>
<p>This is <strong>very</strong> good! We are estimating relative mixtures very well.</p>
</section><section class="slide level2">

<p>Here is the correlation between estimated <code>x</code> and observed <code>x</code>:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1"><span class="kw">cor</span>(<span class="kw">t</span>(est_x), <span class="kw">t</span>(x), <span class="dt">method=</span><span class="st">&#39;spearman&#39;</span>)</a></code></pre></div>
<pre><code>##           [,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]
## [1,] 0.6794778 0.6597489 0.6875596 0.9654864 0.6601344 0.6379879 0.6936902
## [2,] 0.6294459 0.6552449 0.6702929 0.6746403 0.6604643 0.6561367 0.9637838
## [3,] 0.6842969 0.6876411 0.9850094 0.7000887 0.6933407 0.7085986 0.6933872
## [4,] 0.9471063 0.7060617 0.6755484 0.6769583 0.7027036 0.6669958 0.6787414
## [5,] 0.6461627 0.6386977 0.6547104 0.6382085 0.9589411 0.6283625 0.6382968
## [6,] 0.6026265 0.5693803 0.6429443 0.6218183 0.6138628 0.9346406 0.6169345
## [7,] 0.6502485 0.9406994 0.6676147 0.6750623 0.6652670 0.6344473 0.6785245
## [8,] 0.6419115 0.6039992 0.6440546 0.6293570 0.6225426 0.6058642 0.6328787
##           [,8]
## [1,] 0.6886014
## [2,] 0.6930210
## [3,] 0.7292227
## [4,] 0.7381844
## [5,] 0.6623493
## [6,] 0.6557481
## [7,] 0.6609027
## [8,] 0.9395845</code></pre>
<p>This is also amazingly good! We can reconstruct gene expression profiles almost perfectly, despite the substantial overdispersion in expression that we didn’t model using Stan.</p>
</section></section>
<section><section id="t-sne" class="title-slide slide level1"><h1>t-SNE</h1></section><section id="t-sne-1" class="slide level2">
<h2>t-SNE</h2>
<figure>
<img data-src="images/t-sne_abstract.png" alt="http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf" /><figcaption>http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf</figcaption>
</figure>
</section><section class="slide level2">

<p><a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding">t-SNE</a> is a recently published method for dimension reduction (visualization of struture in high-dimensional data).</p>
</section><section class="slide level2">

<p>It <a href="https://distill.pub/2016/misread-tsne/">makes very nice visualizations</a>.</p>
</section><section class="slide level2">

<p>The <em>generic idea</em> is to find a low dimensional representation (i.e., a <em>picture</em>) in which <em>proximity</em> of points reflects <em>similarity</em> of the corresponding (high dimensional) observations.</p>
</section><section class="slide level2">

<p>Suppose we have <span class="math inline">\(n\)</span> data points, <span class="math inline">\(x_1, \ldots, x_n\)</span> and each one has <span class="math inline">\(p\)</span> variables: <span class="math inline">\(x_1 = (x_{11}, \ldots, x_{1p})\)</span>.</p>
<div class="fragment">
<p>For each, we want to find a low-dimensional point <span class="math inline">\(y\)</span>: <span class="math display">\[\begin{aligned}
    x_i \mapsto y_i
\end{aligned}\]</span> similarity of <span class="math inline">\(x_i\)</span> and <span class="math inline">\(x_j\)</span> is (somehow) reflected by proximity of <span class="math inline">\(y_i\)</span> and <span class="math inline">\(y_j\)</span></p>
</div>
<div class="fragment">
<p>We need to define:</p>
<ol start="0" type="1">
<li>“low dimension” (usually, <span class="math inline">\(k=2\)</span> or <span class="math inline">\(3\)</span>)</li>
<li>“similarity” of <span class="math inline">\(x_i\)</span> and <span class="math inline">\(x_j\)</span></li>
<li>“reflected by proximity” of <span class="math inline">\(y_i\)</span> and <span class="math inline">\(y_j\)</span></li>
</ol>
</div>
</section><section id="similarity" class="slide level2">
<h2>Similarity</h2>
<p>To measure similarity, we just use (Euclidean) distance: <span class="math display">\[\begin{aligned}
    d(x_i, x_j) = \sqrt{\sum_{\ell=1}^n (x_{i\ell} - x_{j\ell})^2} ,
\end{aligned}\]</span> after first normalizing variables to vary on comparable scales.</p>
</section><section id="proximity" class="slide level2">
<h2>Proximity</h2>
<p>A natural choice is to require distances in the new space (<span class="math inline">\(d(y_i, y_j)\)</span>) to be as close as possible to distances in the original space (<span class="math inline">\(d(x_i, x_j)\)</span>).</p>
<div class="fragment">
<p>That’s a <em>pairwise</em> condition. t-SNE instead tries to measure how faithfully <em>relative distances</em> are represented in each point’s <em>neighborhood</em>.</p>
</div>
</section><section id="neighborhoods" class="slide level2">
<h2>Neighborhoods</h2>
<p>For each data point <span class="math inline">\(x_i\)</span>, define <span class="math display">\[\begin{aligned}
    p_{ij} = \frac{ e^{-d(x_i, x_j)^2 / (2\sigma^2)} 
                  }{ \sum_{\ell \neq i} e^{-d(x_i, x_\ell)^2 / (2 \sigma^2)} } ,
\end{aligned}\]</span> and <span class="math inline">\(p_{ii} = 0\)</span>.</p>
<p>This is the probability that point <span class="math inline">\(i\)</span> would pick point <span class="math inline">\(j\)</span> as a neighbor if these are chosen according to the Gaussian density centered at <span class="math inline">\(x_i\)</span>.</p>
</section><section class="slide level2">

<p>Similarly, in the output space, define <span class="math display">\[\begin{aligned}
    q_{ij} = \frac{ (1 + d(y_i, y_j)^2)^{-1} 
                  }{ \sum_{\ell \neq i} (1 + d(y_i, y_\ell)^2)^{-1} }
\end{aligned}\]</span> and <span class="math inline">\(q_{ii} = 0\)</span>.</p>
<p>This is the probability that point <span class="math inline">\(i\)</span> would pick point <span class="math inline">\(j\)</span> as a neighbor if these are chosen according to the Cauchy centered at <span class="math inline">\(x_i\)</span>.</p>
</section><section id="similiarity-of-neighborhoods" class="slide level2">
<h2>Similiarity of neighborhoods</h2>
<p>We want neighborhoods to look similar, i.e., choose <span class="math inline">\(y\)</span> so that <span class="math inline">\(q\)</span> looks like <span class="math inline">\(p\)</span>.</p>
<p>To do this, we minimize <span class="math display">\[\begin{aligned}
    \text{KL}(p \given q)
    &amp;=
    \sum_{i} \sum_{j} p_{ij} \log\left(\frac{p_{ij}}{q_{ij}}\right) .
\end{aligned}\]</span></p>
<div class="fragment">
<p><em>What’s KL()?</em></p>
</div>
</section><section id="kullback-leibler-divergence" class="slide level2">
<h2>Kullback-Leibler divergence</h2>
<p><span class="math display">\[\begin{aligned}
    \text{KL}(p \given q)
    &amp;=
    \sum_{i} \sum_{j} p_{ij} \log\left(\frac{p_{ij}}{q_{ij}}\right) .
\end{aligned}\]</span></p>
<p>This is the average log-likelihood ratio between <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> for a single observation drawn from <span class="math inline">\(p\)</span>.</p>
<p>It is a measure of how “surprised” you would be by a sequence of samples from <span class="math inline">\(p\)</span> that you think should be coming from <span class="math inline">\(q\)</span>.</p>
<div class="fragment">
<p><strong>Facts:</strong></p>
<ol type="1">
<li><p><span class="math inline">\(\text{KL}(p \given q) \ge 0\)</span></p></li>
<li><p><span class="math inline">\(\text{KL}(p \given q) = 0\)</span> only if <span class="math inline">\(p_i = q_i\)</span> for all <span class="math inline">\(i\)</span>.</p></li>
</ol>
</div>
</section></section>
<section><section id="simulate-data-1" class="title-slide slide level1"><h1>Simulate data</h1></section><section id="a-high-dimensional-donut" class="slide level2">
<h2>A high-dimensional donut</h2>
<p>Let’s first make some data. This will be some points distributed around a ellipse in <span class="math inline">\(n\)</span> dimensions.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1">n &lt;-<span class="st"> </span><span class="dv">20</span></a>
<a class="sourceLine" id="cb34-2" title="2">npts &lt;-<span class="st"> </span><span class="fl">1e3</span></a>
<a class="sourceLine" id="cb34-3" title="3">xy &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n<span class="op">*</span>npts), <span class="dt">ncol=</span>n)</a>
<a class="sourceLine" id="cb34-4" title="4">theta &lt;-<span class="st"> </span><span class="kw">runif</span>(npts) <span class="op">*</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>pi</a>
<a class="sourceLine" id="cb34-5" title="5">ab &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">2</span><span class="op">*</span>n), <span class="dt">nrow=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb34-6" title="6">ab[,<span class="dv">2</span>] &lt;-<span class="st"> </span>ab[,<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>ab[,<span class="dv">1</span>] <span class="op">*</span><span class="st"> </span><span class="kw">sum</span>(ab[,<span class="dv">1</span>] <span class="op">*</span><span class="st"> </span>ab[,<span class="dv">2</span>]) <span class="op">/</span><span class="st"> </span><span class="kw">sqrt</span>(<span class="kw">sum</span>(ab[,<span class="dv">1</span>]<span class="op">^</span><span class="dv">2</span>))</a>
<a class="sourceLine" id="cb34-7" title="7">ab &lt;-<span class="st"> </span><span class="kw">sweep</span>(ab, <span class="dv">1</span>, <span class="kw">sqrt</span>(<span class="kw">rowSums</span>(ab<span class="op">^</span><span class="dv">2</span>)), <span class="st">&quot;/&quot;</span>)</a>
<a class="sourceLine" id="cb34-8" title="8"><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>npts) {</a>
<a class="sourceLine" id="cb34-9" title="9">    dxy &lt;-<span class="st"> </span><span class="dv">4</span> <span class="op">*</span><span class="st"> </span><span class="kw">c</span>(<span class="kw">cos</span>(theta[k]), <span class="kw">sin</span>(theta[k])) <span class="op">%*%</span><span class="st"> </span>ab</a>
<a class="sourceLine" id="cb34-10" title="10">    xy[k,] &lt;-<span class="st"> </span>xy[k,] <span class="op">+</span><span class="st"> </span>dxy</a>
<a class="sourceLine" id="cb34-11" title="11">}</a></code></pre></div>
</section><section class="slide level2">

<p>Here’s what the data look like: <img src="figure/Week_18_Lecture/plot_data-1.png" title="plot of chunk plot_data" alt="plot of chunk plot_data" style="display: block; margin: auto;" /></p>
</section><section class="slide level2">

<p>But there is hidden, two-dimensional structure: <img src="figure/Week_18_Lecture/plot_ring-1.png" title="plot of chunk plot_ring" alt="plot of chunk plot_ring" style="display: block; margin: auto;" /></p>
</section><section class="slide level2">

<p>Now let’s build the distance matrix.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1">dmat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">dist</span>(xy)) </a></code></pre></div>
</section></section>
<section><section id="implementation" class="title-slide slide level1"><h1>Implementation</h1></section><section id="a-stan-block" class="slide level2">
<h2>A Stan block</h2>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">tsne_block &lt;-<span class="st"> &#39;</span></a>
<a class="sourceLine" id="cb36-2" title="2"><span class="st">data {</span></a>
<a class="sourceLine" id="cb36-3" title="3"><span class="st">    int N; // number of points</span></a>
<a class="sourceLine" id="cb36-4" title="4"><span class="st">    int n; // input dimension</span></a>
<a class="sourceLine" id="cb36-5" title="5"><span class="st">    int k; // output dimension</span></a>
<a class="sourceLine" id="cb36-6" title="6"><span class="st">    matrix[N,N] dsq;  // distance matrix, squared</span></a>
<a class="sourceLine" id="cb36-7" title="7"><span class="st">}</span></a>
<a class="sourceLine" id="cb36-8" title="8"><span class="st">parameters {</span></a>
<a class="sourceLine" id="cb36-9" title="9"><span class="st">    real&lt;lower=0&gt; sigma_sq; // in kernel for p </span></a>
<a class="sourceLine" id="cb36-10" title="10"><span class="st">    matrix[N-2,k] y1;</span></a>
<a class="sourceLine" id="cb36-11" title="11"><span class="st">    real&lt;lower=0&gt; y21;</span></a>
<a class="sourceLine" id="cb36-12" title="12"><span class="st">}</span></a>
<a class="sourceLine" id="cb36-13" title="13"><span class="st">transformed parameters {</span></a>
<a class="sourceLine" id="cb36-14" title="14"><span class="st">    matrix[N,k] y;</span></a>
<a class="sourceLine" id="cb36-15" title="15"><span class="st">    y[1,] = rep_row_vector(0.0, k);</span></a>
<a class="sourceLine" id="cb36-16" title="16"><span class="st">    y[3:N,] = y1;</span></a>
<a class="sourceLine" id="cb36-17" title="17"><span class="st">    y[2,] = rep_row_vector(0.0, k);</span></a>
<a class="sourceLine" id="cb36-18" title="18"><span class="st">    y[2,1] = y21;</span></a>
<a class="sourceLine" id="cb36-19" title="19"><span class="st">}</span></a>
<a class="sourceLine" id="cb36-20" title="20"><span class="st">model {</span></a>
<a class="sourceLine" id="cb36-21" title="21"><span class="st">    matrix[N,N] q;</span></a>
<a class="sourceLine" id="cb36-22" title="22"><span class="st">    real dt;</span></a>
<a class="sourceLine" id="cb36-23" title="23"><span class="st">    matrix[N,N] p;</span></a>
<a class="sourceLine" id="cb36-24" title="24"><span class="st">    q[N,N] = 0.0;</span></a>
<a class="sourceLine" id="cb36-25" title="25"><span class="st">    for (i in 1:(N-1)) {</span></a>
<a class="sourceLine" id="cb36-26" title="26"><span class="st">        q[i,i] = 0.0;</span></a>
<a class="sourceLine" id="cb36-27" title="27"><span class="st">        for (j in (i+1):N) {</span></a>
<a class="sourceLine" id="cb36-28" title="28"><span class="st">            q[i,j] = 1 / (1 + squared_distance(y[i], y[j]));</span></a>
<a class="sourceLine" id="cb36-29" title="29"><span class="st">            q[j,i] = q[i,j];</span></a>
<a class="sourceLine" id="cb36-30" title="30"><span class="st">        }</span></a>
<a class="sourceLine" id="cb36-31" title="31"><span class="st">    }</span></a>
<a class="sourceLine" id="cb36-32" title="32"><span class="st">    for (i in 1:N) {</span></a>
<a class="sourceLine" id="cb36-33" title="33"><span class="st">        q[i] = q[i] / sum(q[i]);</span></a>
<a class="sourceLine" id="cb36-34" title="34"><span class="st">    }</span></a>
<a class="sourceLine" id="cb36-35" title="35"><span class="st">    // create p matrix</span></a>
<a class="sourceLine" id="cb36-36" title="36"><span class="st">    p = exp(-dsq/(2*sigma_sq));</span></a>
<a class="sourceLine" id="cb36-37" title="37"><span class="st">    for (i in 1:N) {</span></a>
<a class="sourceLine" id="cb36-38" title="38"><span class="st">        p[i,i] = 0.0;</span></a>
<a class="sourceLine" id="cb36-39" title="39"><span class="st">        p[i] = p[i] / sum(p[i]);</span></a>
<a class="sourceLine" id="cb36-40" title="40"><span class="st">    }</span></a>
<a class="sourceLine" id="cb36-41" title="41"><span class="st">    // compute the target</span></a>
<a class="sourceLine" id="cb36-42" title="42"><span class="st">    for (i in 1:(N-1)) {</span></a>
<a class="sourceLine" id="cb36-43" title="43"><span class="st">        for (j in (i+1):N) {</span></a>
<a class="sourceLine" id="cb36-44" title="44"><span class="st">            target += (-1) * (p[i,j] .* log(p[i,j] ./ q[i,j]));</span></a>
<a class="sourceLine" id="cb36-45" title="45"><span class="st">            target += (-1) * (p[j,i] .* log(p[j,i] ./ q[j,i]));</span></a>
<a class="sourceLine" id="cb36-46" title="46"><span class="st">        }</span></a>
<a class="sourceLine" id="cb36-47" title="47"><span class="st">    }</span></a>
<a class="sourceLine" id="cb36-48" title="48"><span class="st">    sigma_sq ~ normal(0, 10);</span></a>
<a class="sourceLine" id="cb36-49" title="49"><span class="st">}&#39;</span></a>
<a class="sourceLine" id="cb36-50" title="50"></a>
<a class="sourceLine" id="cb36-51" title="51"></a>
<a class="sourceLine" id="cb36-52" title="52">tk &lt;-<span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb36-53" title="53">tsne_model &lt;-<span class="st"> </span><span class="kw">stan_model</span>(<span class="dt">model_code=</span>tsne_block)</a></code></pre></div>
</section><section id="run-the-model-with-optimizing" class="slide level2">
<h2>Run the model with <code>optimizing</code></h2>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1">runtime &lt;-<span class="st"> </span><span class="kw">system.time</span>(tsne_optim &lt;-<span class="st"> </span><span class="kw">optimizing</span>(tsne_model,</a>
<a class="sourceLine" id="cb37-2" title="2">                                 <span class="dt">data=</span><span class="kw">list</span>(<span class="dt">N=</span><span class="kw">nrow</span>(xy),</a>
<a class="sourceLine" id="cb37-3" title="3">                                           <span class="dt">n=</span><span class="kw">ncol</span>(xy),</a>
<a class="sourceLine" id="cb37-4" title="4">                                           <span class="dt">k=</span>tk,</a>
<a class="sourceLine" id="cb37-5" title="5">                                           <span class="dt">dsq=</span>(dmat<span class="op">/</span><span class="kw">max</span>(dmat))<span class="op">^</span><span class="dv">2</span>)))</a>
<a class="sourceLine" id="cb37-6" title="6">runtime</a></code></pre></div>
<pre><code>##    user  system elapsed 
## 117.588   0.180 117.799</code></pre>
</section><section id="it-works" class="slide level2">
<h2>It works!</h2>
<p><img src="figure/Week_18_Lecture/plot_tsne-1.png" title="plot of chunk plot_tsne" alt="plot of chunk plot_tsne" style="display: block; margin: auto;" /></p>
</section></section>
<section><section id="another-case" class="title-slide slide level1"><h1>Another case</h1></section><section id="high-dimensional-random-walk" class="slide level2">
<h2>High-dimensional random walk</h2>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1">n &lt;-<span class="st"> </span><span class="dv">40</span></a>
<a class="sourceLine" id="cb39-2" title="2">npts &lt;-<span class="st"> </span><span class="fl">1e2</span></a>
<a class="sourceLine" id="cb39-3" title="3"><span class="co"># rw &lt;- colCumsums(matrix(rnorm(n*npts), ncol=n))</span></a>
<a class="sourceLine" id="cb39-4" title="4">rw &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n<span class="op">*</span>npts), <span class="dt">ncol=</span>n)</a>
<a class="sourceLine" id="cb39-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span>npts) {</a>
<a class="sourceLine" id="cb39-6" title="6">    rw[i,] &lt;-<span class="st"> </span><span class="kw">sqrt</span>(.<span class="dv">9</span>) <span class="op">*</span><span class="st"> </span>rw[i<span class="dv">-1</span>,] <span class="op">+</span><span class="st"> </span><span class="kw">sqrt</span>(.<span class="dv">1</span>) <span class="op">*</span><span class="st"> </span>rw[i,]</a>
<a class="sourceLine" id="cb39-7" title="7">}</a></code></pre></div>
</section><section class="slide level2">

<p>Here’s what the data look like: <img src="figure/Week_18_Lecture/plot_rw-1.png" title="plot of chunk plot_rw" alt="plot of chunk plot_rw" style="display: block; margin: auto;" /></p>
</section><section class="slide level2">

<p><img src="figure/Week_18_Lecture/plot_rw_sub-1.png" title="plot of chunk plot_rw_sub" alt="plot of chunk plot_rw_sub" style="display: block; margin: auto;" /></p>
</section><section class="slide level2">

<p>Now let’s build the distance matrix.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" title="1">rw_dmat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">dist</span>(rw)) </a></code></pre></div>
</section><section id="run-again" class="slide level2">
<h2>Run again</h2>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" title="1">rw_runtime &lt;-<span class="st"> </span><span class="kw">system.time</span>(</a>
<a class="sourceLine" id="cb41-2" title="2">     rw_tsne_optim &lt;-<span class="st"> </span><span class="kw">optimizing</span>(tsne_model,</a>
<a class="sourceLine" id="cb41-3" title="3">                                 <span class="dt">data=</span><span class="kw">list</span>(<span class="dt">N=</span><span class="kw">nrow</span>(rw),</a>
<a class="sourceLine" id="cb41-4" title="4">                                           <span class="dt">n=</span><span class="kw">ncol</span>(rw),</a>
<a class="sourceLine" id="cb41-5" title="5">                                           <span class="dt">k=</span>tk,</a>
<a class="sourceLine" id="cb41-6" title="6">                                           <span class="dt">dsq=</span>(rw_dmat<span class="op">/</span><span class="kw">max</span>(rw_dmat))<span class="op">^</span><span class="dv">2</span>),</a>
<a class="sourceLine" id="cb41-7" title="7">                                 <span class="dt">tol_rel_grad=</span><span class="fl">1e5</span>))</a>
<a class="sourceLine" id="cb41-8" title="8">rw_runtime</a></code></pre></div>
<pre><code>##    user  system elapsed 
##   1.038   0.000   1.038</code></pre>
</section><section id="it-works-1" class="slide level2">
<h2>It works!</h2>
<p><img src="figure/Week_18_Lecture/plot_rw_tsne-1.png" title="plot of chunk plot_rw_tsne" alt="plot of chunk plot_rw_tsne" style="display: block; margin: auto;" /></p>
</section></section>
    </div>
  </div>

  <script src="reveal.js/lib/js/head.min.js"></script>
  <script src="reveal.js/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // Optional reveal.js plugins
        dependencies: [
          { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true },
          { src: 'reveal.js/plugin/math/math.js', async: true },
          { src: 'reveal.js/plugin/notes/notes.js', async: true }
        ]
      });
    </script>
    </body>
</html>
